packages/react-router-dom/index.tsx

```diff
@@ -1220,21 +1230,33 @@ function useScrollRestoration({
     React.useCallback(() => {
       if (navigation.state === "idle") {
         let key = (getKey ? getKey(location, matches) : null) || location.key;
-        savedScrollPositions[key] = getScrollY(elementRef);
+        savedScrollPositions[key] = getScrollY(scrollContainer);
       }
       sessionStorage.setItem(
         storageKey || SCROLL_RESTORATION_STORAGE_KEY,
         JSON.stringify(savedScrollPositions)
       );
       window.history.scrollRestoration = "auto";
-    }, [storageKey, getKey, elementRef, navigation.state, location, matches])
+    }, [
+      storageKey,
+      getKey,
+      scrollContainer,
+      navigation.state,
+      location,
+      matches,
+    ])
   );
 
   // Read in any saved scroll locations
   if (typeof document !== "undefined") {
     // eslint-disable-next-line react-hooks/rules-of-hooks
     React.useLayoutEffect(() => {
       try {
+        // Only load once per session so it persists between unmounts/remounts
+        // (i.e., different instance per layout)
+        if (savedScrollPositions) {
+          return;
+        }
```

Comment:
Ensure in-memory positions are maintained if we unmount/remount a new `<ScrollRestoration>`

Reply:
FWIW, it doesn't look like `savedScrollPositions` is ever allowed to be falsy, as it's initialized to `{}`.